### ISSUE TITLE
blank webui.source page  without any error

### ISSUE THREAD

**Issue by fanlix** [google/pprof#621]
# blank webui.source page with self built pprof, but go tool pprof works.
 
### What version of pprof are you using?
* commit 94a9f03dee38882adc8bdfc42cdd6a04f8e7056e 

### What operating system and processor architecture are you using?
* ubuntu Linux  5.4.0-65-generic #73-Ubuntu SMP Mon Jan 18 17:25:17 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
* go1.15.5 linux/amd64

### What did you do?
* build
  * go version go1.15.5 linux/amd64
  * `go build`
* generate profile data file, with src file
  * [pprof-data-t3.zip](https://github.com/google/pprof/files/6276149/pprof-data-t3.zip) 
* run `./pprof -http :9999 t3.pf`

### What did you expect to see?
webui source page

### What did you see instead?
* top, graph, ...  other pages work fine
* blank source page, no error, no log

### `go tool pprof` works
`go tool pprof -http :9999 t3.pf`
* go version go1.15.5 linux/amd64
* go version go1.16.2 linux/amd64

### build for different OS
the same, blank source page
* linux
* windows 

different graphviz version, same.
* graphviz 2.26
* graphviz 2.40
* graphviz 2.47

### -source_path option
* `./pprof -http :9999  -source_path=.  t3.pf`

## Question
* is there any build flags should i put when `go build` ? (to build as go tool version)
* is there any secret pprof cmdline options to enable verbose output for debug?
  * ./pprof -v
* how to know binary pprof version?  (to build as go tool version)
  * `go tool pprof -version`

**Comment by fanlix** [comment]
## update:
try build from old version: 
* such as 2019 commit d4f498aebedc02388a1c34ebe8dcb2e7c9e99f29

webui.source page works fine.

**Comment by wyk9787** [comment]
Thanks for reporting the issue and I really appreciate various options you have tried. Yes, this is indeed an issue and I am able to reproduce it.

I think the culprit is https://github.com/google/pprof/pull/610 (the commit before it doesn't have this issue), hence asking @gmarin13 to take a look.

Regarding your three questions, asking @nolanmar511 to supplement here, but my answers are "no" here unfortunately:
1. I am not aware if there is a magic flag you can pass here, but I think there might be a way to find which pprof commit Golang uses to include pprof into its release. @nolanmar511, would you happen to know where we can find that?
2. No, as far as I know.
3. No, as far as I know.

Wish I can be more helpful here, and thanks again for reporting the issue.

**Comment by gmarin13** [comment]
Thanks for reporting. I would like to reproduce, but I am not as versed in pprof command line options.
How can I get to that source listing without starting a local http server, which I cannot do?

Also, #620 effectively disables the changes in #610 and most of the changes in #598, so I don't think #610 can be the culprit. I would consider #598 more likely to break things, but as I said, #620 disables that code for now.

@wyk9787 is it possible to reproduce this without using a local http server?

**Comment by nolanmar511** [comment]
The pprof commit used by go1.16.2 is https://github.com/google/pprof/commit/1bf35d6f28c2d0288b382023b376d4db634119e2

This commit can be found here in GitHub repo for go: https://github.com/golang/go/blob/3979fb9af9ccfc0b7ccb613dcf256b18c2c295f0/src/cmd/go.mod#L6

**Comment by gmarin13** [comment]
I am able to reproduce. I am looking into it.

**Comment by gmarin13** [comment]
In my testing, source listing is broken at commit [cbba55b83ad56c9286566b96b0c82cead332b729](https://github.com/google/pprof/commit/cbba55b83ad56c9286566b96b0c82cead332b729).

Steps to reproduce:
# In my local pprof repo.
$ git checkout -b fix_source cbba55b83ad56c9
$ go build
$ pprof -http :9999 t3.pf 
Serving web UI on http://localhost:9999

-> Source listing is empty.

# Testing at the previous commit [c5db671c31d6a900d3dae7fc37f73ea7e4c68e5c](https://github.com/google/pprof/commit/c5db671c31d6a900d3dae7fc37f73ea7e4c68e5c)
$ git reset --hard c5db671c31d6a900d
$ go build
$ pprof -http :9999 t3.pf 
Serving web UI on http://localhost:9999

-> Source listing is showing.

List of commits in chronological order at https://github.com/google/pprof/commits/master

**Comment by gmarin13** [comment]
@ghemawat FYI.

**Comment by ghemawat** [comment]
When I look at the profile (with pprof -raw), I see that all of the locations listed in the profile have an address of zero (see below). How was this profile generated? Perhaps it is an artifact of how lua code is profiled?

The fact that there are no real addresses in the profile explains the current behavior of the new source listing code. Question for Alex and others on the pprof team. What is the specified interpretation of Address=0 in location. Are we supposed to support such profiles? perhaps by synthesizing fake addresses?

Even if we fix that, the other problem is that the source listing code calls objectTool.Disasm. But there doesn't seem to be a binary.  Should we consider falling back to the embedded line number in the profile location entries for such cases?

FYI, here is the pprof -raw output:

```
Comment: lua-503 pprof, tsv-0.9.4
PeriodType:
Period: 0
Time: 2021-04-07 20:34:07 -0700 PDT
Duration: 1.23
Samples:
samples/count cpu/nanoseconds[dflt] alloc_count/count alloc_total/bytes alloc_inuse/bytes
          1   60695073          0          0          0: 1 2 2
         10     101270          0          0          0: 3 4 2 2
          0          0         20        800        800: 5 6 4 2 2
         10      41705          0          0          0: 7 6 4 2 2
          0          0         20       8560       8560: 7 7 6 4 2 2
         30     353608          0          0          0: 8 4 2 2
          0          0         30       1020       1020: 8 8 4 2 2
          0          0         30       1680       1680: 9 4 2 2
          0          0        330     982560     491520: 10 4 2 2
          0          0     102350   43751200    9767154: 11 4 2 2
      30000   57416646          0          0          0: 12 4 2 2
          0          0      77389   42826972  -10474843: 12 12 4 2 2
          0          0        252     419328     209664: 13 4 2 2
         30      85481          0          0          0: 14 4 2 2
          0          0         30       2160       2160: 14 15 4 2 2
         30       4341          0          0          0: 16 15 4 2 2
         30      40523          0          0          0: 17 15 4 2 2
          0          0         10        720        720: 17 17 15 4 2 2
         10          0          0          0          0: 18 2 2
          0          0         10        720        720: 18 19 2 2
         10       7818          0          0          0: 20 19 2 2
         10          0         10      10880       5440: 21 19 2 2
         30       6890          0          0          0: 22 23 19 2 2
         30     140201          0          0          0: 24 23 19 2 2
          0          0        120       4320       4320: 24 24 23 19 2 2
         30      32197          0          0          0: 25 23 19 2 2
          0          0         30        870        870: 25 25 23 19 2 2
         30      69786          0          0          0: 12 23 19 2 2
          0          0         30       1080       1080: 12 12 23 19 2 2
         30   19939182          0          0          0: 26 23 19 2 2
         30   59335348          0          0          0: 27 23 19 2 2
          0          0         10        720        720: 27 28 23 19 2 2
     379976 1081166470          0          0          0: 29 28 23 19 2 2
         30     105923          0          0          0: 30 23 19 2 2
         30       5279          0          0          0: 31 30 23 19 2 2
         30     198385          0          0          0: 32 30 23 19 2 2
          1          0          0          0          0: 33 2 2
Locations
     1: 0x0 f_loop t3.lua:19 s=18()
     2: 0x0 root root.src:0 s=1()
     3: 0x0 create@.\profile.lua .\profile.lua:28 s=27()
     4: 0x0 f_loop t3.lua:21 s=18()
     5: 0x0 create@.\profile.lua .\profile.lua:31 s=27()
     6: 0x0 create@.\profile.lua .\profile.lua:21 s=27()
     7: 0x0 real_co_create c:src:0 s=0()
     8: 0x0 status c:src:0 s=0()
     9: 0x0 f_loop t3.lua:24 s=18()
    10: 0x0 f_loop t3.lua:28 s=18()
    11: 0x0 f_loop t3.lua:29 s=18()
    12: 0x0 format c:src:0 s=0()
    13: 0x0 f_loop t3.lua:30 s=18()
    14: 0x0 resume@.\profile.lua .\profile.lua:44 s=43()
    15: 0x0 resume@.\profile.lua .\profile.lua:35 s=43()
    16: 0x0 hookresume@.\profile.lua .\profile.lua:21 s=21()
    17: 0x0 real_co_resume c:src:0 s=0()
    18: 0x0 null@.\profile.lua .\profile.lua:29 s=28()
    19: 0x0 null@.\profile.lua .\profile.lua:0 s=28()
    20: 0x0 hookstart c:src:0 s=0()
    21: 0x0 nil t3.lua:8 s=7()
    22: 0x0 running c:src:0 s=0()
    23: 0x0 nil t3.lua:11 s=7()
    24: 0x0 tostring c:src:0 s=0()
    25: 0x0 sub c:src:0 s=0()
    26: 0x0 write c:src:0 s=0()
    27: 0x0 sleep t3.lua:3 s=2()
    28: 0x0 sleep t3.lua:12 s=2()
    29: 0x0 clock c:src:0 s=0()
    30: 0x0 yield@.\profile.lua .\profile.lua:40 s=39()
    31: 0x0 hookyield@.\profile.lua .\profile.lua:20 s=20()
    32: 0x0 real_co_yield c:src:0 s=0()
    33: 0x0 stop c:src:0 s=0()
Mappings
1: 0x0/0x0/0x0 @t3.lua

```

**Comment by fanlix** [comment]
@gmarin13 yes, there are no function address, no mapping, no binary exe in this profile data, which generated by my lua-profile-tool. It works fine with `go tool pprof` for a while.

`pprof` plays a role as data parsing and ui displaying in my tool chain. 
I used to design a file format myself to store profile data, and generate svg/webui by hand. This is usable, but hard for maintenance.
After learned golang, I'm quite satisfied with pprof. Then i migrate my tool to works together with pprof. Perfect! 

Lua functions has address, but it is **runtime dynamic**. 
* we cannot get asm code from binary file with (func.addr, objexe).
  * C part: yes
  * Lua part: no
* but we can get the source code from (source.filename, line).

Lua may compile source code to binary opcode, something like asm.
Maybe i should link pprof's disassemble page to lua opcode, but it is tons of works.
So for now, disassemble page is unusable for my profile data.

**Comment by ghemawat** [comment]
@fanlix - a suggestion: when you generate the profile for your Lua code, can you give every location a distinct address? (Right now they are all zero.)  The address does not need to mean anything, but it might be good to avoid addresses that are in C code. E.g., in the future you might want to merge with a profile of C code to get a more complete view.

**Comment by aalexand** [comment]
I believe Java profiles also do not include an address. Perhaps Python ones
do not either. Not sure we should require producers of profiles for
interpreted (or JIT - anything where capturing the runtime address is not
possible, hard or not useful) language implementations to include
artificial addresses.

**Comment by ghemawat** [comment]
Understood. My current PR does not require locations to have addresses. Does somebody want to review/merge it in?

**Comment by fanlix** [comment]
As suggested, I generate the profile data with function addresses, obj mappings.
### this is the merge result of lua/clua/c/vm, **complete view**.
* pure lua functions keep the same
* clua functions written in c code, now show as c functions, with detail call stack and c source file
* lua vm engine functions show as c function.

but, for developers not familier with lua vm and src, it is unreadable. 
* All funcitons in lua code are shown as low level c functions
* One easy api call is expended to a long c call stack


We commonly trust lua vm engine and most libs.
For lua clib developers, this merge view maybe useful. For end users, it's better to keep profiling at lua level.

**Comment by zikaeroh** [comment]
One thing to note is that the version of pprof that's currently slated to be included with Go 1.17 has this bug; testing `gotip tool pprof` on my Go profiles always gives me an empty source page.

**Comment by ghemawat** [comment]
Hi Zik,

Thanks. The problem was because the new source listing code was dependent on having the binary available.

I pushed an updated PR that fixes the problem. Do you know who to contact to pick up the PR (when it has been merged) in case Go 1.17 has already been frozen? If not, that's fine, I can ask around and figure it out.

**Comment by zikaeroh** [comment]
I filed https://github.com/golang/go/issues/45786 to ensure there's an issue on the Go repo for consideration. Since this is a bugfix, I'd hope they'd accept the change even if the PR isn't merged before the code freeze this week.
